drivers: can: Add CAN driver support

alexanderwachter opened the pull request:
I created a driver API for Controller Area Network. Also a low level driver for stm32 micro controllers is already implemented. I tested it with the stm32f072b disco board in loopback mode and with two boards connected via additional can  drivers.

Commit:
samples: can: Add example code for CAN driver

This adds example and testing code for CAN driver.
Tested on stm32f072b disco.
Examples are given for:
- can_configure
- can_attach_isr
- can_attach_msgq
- can_send

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

jukkar commented on boards/arm/stm32f072b_disco/pinmux.c:
```diff
@@ -32,6 +32,11 @@ static const struct pin_config pinconf[] = {
 	{STM32_PIN_PB5, STM32F0_PINMUX_FUNC_PB5_SPI1_MOSI},
 #endif /* CONFIG_SPI_1 */
+#ifdef CONFIG_CAN_1
+   {STM32_PIN_PB8, STM32F0_PINMUX_FUNC_PB8_CAN_RX},
```
Indentation seems to be wrong here

alexanderwachter replied:
Sorry for this huge bunch of styling issues. I should have checked this before submitting. Shame on me :-)

jukkar commented on boards/arm/stm32f072b_disco/stm32f072b_disco.dts:
```diff
@@ -74,3 +74,10 @@
+
+&can1 {
+	pinctrl-0 = <&can_pins_a>;
+    pinctrl-names = "default";
```
here too

jukkar commented on boards/arm/stm32f072b_disco/stm32f072b_disco_defconfig:
```diff
@@ -27,8 +27,14 @@ CONFIG_GPIO=y
 CONFIG_CLOCK_STM32_APB1_PRESCALER=1
+# produce 48MHz clock at PLL output
+CONFIG_CLOCK_STM32_PLL_PREDIV1=1
+
```
empty lines at the end of file should be removed

jukkar commented on drivers/can/Kconfig:
```diff
@@ -0,0 +1,91 @@
+config CAN_PHASE_SEG1_PROP_SEG
+	int "Phase_Seg1_Prop_Seg"
+	default 5
+    range 1 16
```
Indentation is wrong for the "range" option here and below

jukkar commented on drivers/can/Kconfig:
```diff
@@ -0,0 +1,91 @@
+	help
+	  Enable CAN controller 1
+
+config CAN_STM32
```
I think you should not have SMT32 specific stuff in this generic driver.

alexanderwachter replied:
do you mean the config CAN_MAX_FILTER?
config CAN_STM32 enables the STM32 driver. I did it like the i2c driver.
Should I make a own file for this two parameters?

jukkar replied:
It just looked wrong to have CAN_STM32 in this file, usually board specific stuff is in its own Kconfig file.

alexanderwachter replied:
Ok. I'll add a Kconfig.stm32

jukkar commented on drivers/can/stm32_ll_can.c:
```diff
@@ -0,0 +1,725 @@
+	msg->dlc = mbox->RDTR & CAN_RDT0R_DLC >> CAN_RDT0R_DLC_Pos;
+	msg->data_32[0] = mbox->RDLR;
+	msg->data_32[1] = mbox->RDHR;
+
```
remove empty line here

jukkar commented on drivers/can/stm32_ll_can.c:
```diff
@@ -0,0 +1,725 @@
+		msg->std_id =  mbox->RIR >> CAN_RI0R_STID_Pos;
+		msg->id_type = can_standard_identifier;
+	}
+	msg->rtr = mbox->RIR & CAN_RI0R_RTR ? can_remoterequest : can_dataframe;
```
put empty line after }

jukkar commented on drivers/can/stm32_ll_can.c:
```diff
@@ -0,0 +1,725 @@
+	if (mbox->RIR & CAN_RI0R_IDE) {
+		msg->ext_id = mbox->RIR >> CAN_RI0R_EXID_Pos;
+		msg->id_type = can_extended_identifier;
+	} else   {
```
remove extra spaces after else

alexanderwachter replied:
I used uncrustify for this file and it changed it from if \n {} \n else.. to this. I will change that.

jukkar commented on drivers/can/stm32_ll_can.c:
```diff
@@ -0,0 +1,725 @@
+		CAN_FIFOMailBox_TypeDef *mbox = &can->sFIFOMailBox[0];
+		int filter_match_index = ((mbox->RDTR & CAN_RDT0R_FMI_Msk)
+					  >> CAN_RDT0R_FMI_Pos);
+		if (filter_match_index >= CONFIG_CAN_MAX_FILTER) {
```
Put empty line after variable declarations

jukkar commented on drivers/can/stm32_ll_can.c:
```diff
@@ -0,0 +1,725 @@
+		if (filter_match_index >= CONFIG_CAN_MAX_FILTER) {
+			break;
+		}
+		SYS_LOG_DBG("Message on filter index %d\n", filter_match_index);
```
put empty line after }

jukkar commented on drivers/can/stm32_ll_can.c:
```diff
@@ -0,0 +1,725 @@
+
+void HAL_CAN_MspInit(CAN_HandleTypeDef *hcan)
+{
+
```
ARG_UNUSED(hcan); is missing

jukkar commented on drivers/can/stm32_ll_can.c:
```diff
@@ -0,0 +1,725 @@
+	if (!bitrate) {
+		bitrate = cfg->bus_speed;
+	}
+	u32_t prescaler = clock_rate / (BIT_SEG_LENGTH * cfg->bus_speed);
```
No variable declarations in the middle of the function

jukkar commented on drivers/can/stm32_ll_can.c:
```diff
@@ -0,0 +1,725 @@
+	struct k_mutex *tx_mutex = &data->tx_mutex;
+	struct can_mailbox *mb = NULL;
+
+	SYS_LOG_DBG("Sending %d bytes on %s\n"
```
Normally syslog prints should not have \n in them as that will look funny if the syslog is processed remotely. I think you could condense the line and have everything in one line. Note that syslog will automatically place \n at the end of the printed line if printing to console so you do not have to put it there.

jukkar commented on drivers/can/stm32_ll_can.c:
```diff
@@ -0,0 +1,725 @@
+		k_mutex_lock(tx_mutex, K_FOREVER);
+		transmit_status_register = can->TSR;
+	}
+	if (transmit_status_register & CAN_TSR_TME0) {
```
empty line after }

jukkar commented on drivers/can/stm32_ll_can.c:
```diff
@@ -0,0 +1,725 @@
+		mailbox = &can->sTxMailBox[CAN_TXMAILBOX_2];
+		mb = &data->mb2;
+	}
+	mb->tx_callback = callback;
```
same here

jukkar commented on include/can.h:
```diff
@@ -0,0 +1,369 @@
+ * identifier
+ */
+typedef enum {
+	can_standard_identifier,
```
Usually enums / defines are written in all capital letters

jukkar commented on include/can.h:
```diff
@@ -0,0 +1,369 @@
+	/** Indicates the identifier type (standard or extended)
+	 * use can_ide_t for assignment
+	 */
+	u32_t id_type : 1;
```
Using bitfields in structs that are received from wire is problematic as this is compiler specific stuff how it places the bitfields. I think better would be to not to use them here but have access functions that get/set those bits instead.

alexanderwachter replied:
Bitfields are here used to save memory. They are anyway not written directly form the controller registers but set with the enum.

jukkar replied:
Ok. It just looked that this struct is directly mapped to bytes received from wire, especially as there was the __packed attribute there (which is usually not used for normal memory only structs).

jukkar commented on samples/can/src/main.c:
```diff
@@ -0,0 +1,294 @@
+	msg.id_type = can_extended_identifier;
+	msg.dlc = 0;
+	msg.rtr = can_dataframe;
+	for(int str_len = strlen(string); str_len;)
```
do "for (str_len = ...." instead here, so space after "for" and place variable declaration at the start of the func

jukkar commented on samples/can/src/main.c:
```diff
@@ -0,0 +1,294 @@
+		string += msg.dlc;
+		can_send(can_dev, &msg, 10, tx_irq_callback);
+	}
+	
```
remove empty line

jukkar commented on samples/can/src/main.c:
```diff
@@ -0,0 +1,294 @@
+
+	printk("TX thread is running.\n");
+	while (1)
+	{
```
We do not write loops like this, do "while (1) {" instead

jukkar commented on samples/can/src/main.c:
```diff
@@ -0,0 +1,294 @@
+#include <gpio.h>
+
+#define CAN_DEV "CAN_1"
+#define GPIO_LED_DEV "GPIOC"
```
You could place all these board specific options to sample specific Kconfig file.

alexanderwachter replied:
do you rally think this is a good idea?
In my opinion samples are not configured by Kconfig.
Can you give me an example which sample code uses this?

jukkar replied:
See for example https://github.com/zephyrproject-rtos/zephyr/tree/master/samples/net/stats
My point was to put board specific stuff in sample app to Kconfig so that user does not have to change the actual code if he wants to run in different board where the settings might be different.
You need to also add `set(KCONFIG_ROOT ${CMAKE_CURRENT_SOURCE_DIR}/Kconfig)` to CMakeLists.txt for this to work.

alexanderwachter replied:
Sounds reasonable. I'll do that.

jukkar commented on samples/can/src/main.c:
```diff
@@ -0,0 +1,294 @@
+void rx_str_thread(void *msgq, void *can_dev_param, void *unused)
+{
+	struct can_msg msg;
+	{
```
having {} here looks unnecessary.

alexanderwachter replied:
This is to create a scope for the filter. I wanted to show that the filter struct can be discarded after the attach function.

tbursztyka replied:
well, document can_attach function in the api about that behavior then (if not already).

This kind of lone block is a bit ugly

jukkar commented on samples/can/src/main.c:
```diff
@@ -0,0 +1,294 @@
+	}
+
+	int ret = gpio_pin_configure(button_gpio_dev, PIN_USER_BUTTON,
+								 (GPIO_DIR_IN | GPIO_INT |
```
Indentation wrong

jukkar commented on samples/can/src/main.c:
```diff
@@ -0,0 +1,294 @@
+								  GPIO_INT_EDGE | GPIO_INT_ACTIVE_HIGH |
+								  GPIO_INT_DEBOUNCE));
+
+	if (ret)
```
missing {}

jukkar reviewed (changes requested):
Look quite decent to me. There was lot of style issues, you could use checkpatch.pl and uncrustify to check/fix these in your code.

jukkar commented on drivers/can/stm32_ll_can.c:
```diff
@@ -0,0 +1,752 @@
+		return -EINVAL;
+	}
+
+	const u32_t bs1 = (CONFIG_CAN_PHASE_SEG1_PROP_SEG - 1) << CAN_BTR_TS1_Pos;
```
Sometimes it is tempting thing to do, but in zephyr we try to avoid declaring variables in the middle of the function

jukkar commented on drivers/can/stm32_ll_can.c:
```diff
@@ -0,0 +1,752 @@
+		SYS_LOG_ERR("HAL_CAN_Init failed: %d", hal_ret);
+		return -EIO;
+	}
+	SYS_LOG_DBG("Runtime configure of %s done", dev->config->name);
```
Put empty line after }

jukkar commented on drivers/can/stm32_ll_can.c:
```diff
@@ -0,0 +1,752 @@
+	SYS_LOG_DBG("Detatch filter number %d (index %d), type %d", filter_nr,
+		    filter_index,
+		    type);
+	u32_t reset_mask = (type == CAN_FILTER_STANDARD) ? 0x01 :
```
Declare variable at the beginning of func

jukkar commented on drivers/can/stm32_ll_can.c:
```diff
@@ -0,0 +1,752 @@
+	} else {
+		SYS_LOG_DBG("Bank number %d is empty -> deakivate", bank_nr);
+	}
+	can->FMR &= ~(CAN_FMR_FINIT);
```
empty line after }

jukkar commented on drivers/can/stm32_ll_can.c:
```diff
@@ -0,0 +1,752 @@
+	u32_t scale_reg;
+	can_filter_type_t type;
+
+	k_mutex_lock(&data->set_filter_mutex, K_FOREVER);
```
Next 20 lines are quite tightly packed, could you add some empty lines there so it would look a bit nicer to reader.

jukkar reviewed (commented):
Some minor comments. Someone else needs to review this too as this is a new API proposal.

erwango reviewed (changes requested):
Thanks for this work!
Some comments

alexanderwachter commented on boards/arm/stm32f072b_disco/stm32f072b_disco_defconfig:
```diff
@@ -27,8 +27,12 @@ CONFIG_GPIO=y
+CONFIG_CLOCK_STM32_HSE_CLOCK=8000000
 # PLL configuration
-CONFIG_CLOCK_STM32_PLL_SRC_HSI=y
+CONFIG_CLOCK_STM32_PLL_SRC_HSE=y
```
Sorry, I forgot to change this back. I used the HSE for a more precise clock. I didn't test CAN with the internal oscillator, but I will do that.

alexanderwachter replied:
I tested the HSI configuration in loopback mode and in normal mode with a second board using HSE connected. The HSI seems to be precise enough.
Thanks for the hint.

alexanderwachter commented on drivers/can/Kconfig.stm32:
```diff
@@ -0,0 +1,25 @@
+	default n
+	select USE_STM32_HAL_CAN
+	help
+	  Enable STM32 CAN Driver
```
I have a NUCLEO-F746ZG too. I will check it there as soon as the board support on zephyr is ready.

psidhu commented on include/can.h:
```diff
@@ -0,0 +1,380 @@
+	CAN_LOOPBACK_MODE,
+	/*Combination of loopback and silent*/
+	CAN_SILENT_LOOPBACK_MODE
+} can_mode_t;
```
Please don't typedef these enumerations. This makes it unnecessarily confusing when using public APIs.

Instead, your ``can_configure(struct device *dev, can_mode_t mode, u32_t bitrate);`` would become ``can_configure(struct device *dev, enum can_mode mode, u32_t bitrate);`` which makes it painfully obvious to anyone writing / looking at a driver what the value type is being passed around.

See: 
* http://docs.zephyrproject.org/contribute/contribute_guidelines.html#coding-style
* https://www.kernel.org/doc/html/latest/process/coding-style.html#typedefs

alexanderwachter replied:
Thanks for the hint. I changed that.

erwango commented:
@alexanderwachter , looking you PR more closely, it appears that you're not using STM32Cube LL API (STM32Cube LL flavor API not being available for CAN driver as of today). Besides, a part from Init, you're not using HAL API neither.
So, in order to keep consistency in STM32 drivers naming in Zephyr, please rename you driver as stm32_can.c and stm32_can.h.

Also, it would be great to split a bit more the commits:
-API
-Driver/ Kconfig
-Device tree / Bindings
-Board
-Sample

Thanks
Erwan

alexanderwachter commented:
@erwango in the first versions I tried to use STM32Cube but sadly there is only a HAL driver but no LL driver. I decided then not to use it because I can't use HAL_Init which changes the Timers.
Also the overhead of copying messages around is removed by not using the HAL driver.
Anyway setting the registers is not the major part but managing threads and filters in a way that the user does not has to care was the harder part ;-). But my work was heavily inspired by the STM32Cube driver.

erwango commented:
@alexanderwachter , thanks for this explanation. Indeed having LL would be be nice, I'll check if there are some plans to provide them.
Anyway, please consider changing file names for consistency with other STM32 drivers.

alexanderwachter commented:
@erwango An automated filter management as I made with can_stm32_set_filter would be nice too ;)

alexanderwachter commented:
@erwango does it make sense to commit the include/can.h without the can_handlers.c? Because if I want to commit the can_handlers I also have to commit the CMakeLists and because of the add_subdirectory_if_kconfig I also need the Kconfig.

Commit:
API: can: Add API for Controller Area Network driver

This API defines following calls

- can_configure
- can_send
- can_attach_isr
- can_attach_msgq
- can_detach

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

Commit:
drivers: can: Add Kconfig for CAN driver

This commit adds the Kconfig for generic CAN drivers.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

Commit:
drivers: can: Add syscall handlers for CAN API

This commit adds generic syscall handlers for the CAN API

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

Commit:
drivers: can: Add dts bindings for CAN

This commit adds generic device tree bindings for can devices.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

Commit:
drivers: can: Add CAN driver support for STM32

This commit adds low level driver support for STM32 micro controllers.
It is tested on stm32f072b in loopback and normal mode.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

Commit:
boards: stm: Add CAN support for stm32f072b micro controller

This commit adds the can device to the stm32f072b device tree and pinmux

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

erwango commented:
@alexanderwachter Ok , then keep Kconfig along with API introduction

alexanderwachter commented:
I decided to push the can.h extra because this commit is probably the most critical one.
Thank you for the suggestions.
Is it ok now?

alexanderwachter commented:
Any other suggestions regarding the API? Do we need more than 
- configure
- attach_isr
- attach_msgq
- detach
- send?

What about the parameters of configure? 
In principle we can configure more than mode and speed but those things I left in the device tree.
What about attach_msgq?
I thought it is a good idea to have a built in function to offload work and the effort in low level drivers is not really big.

tbursztyka commented on include/can.h:
```diff
@@ -0,0 +1,380 @@
+ * @retval -EIO General input / output error, failed to configure device.
+ */
+typedef int (*can_configure_t)(struct device *dev, enum can_mode mode,
+							   u32_t bitrate);
```
wrong parameter alignment: it should be aligned below previous '(' + 1 space.

Apply that comment everywhere, it's the same when calling functions etc...

alexanderwachter replied:
ohh. Now I got it. I was wondering why it always looked so bad after submitting. I used 4 spaces tabs instead of 8!

tbursztyka commented on include/can.h:
```diff
@@ -0,0 +1,380 @@
+ *
+ * @retval none
+ */
+typedef void (*can_detach_t)(struct device *dev, int filter_nr);
```
s/filter_nr/filter_id ?

alexanderwachter replied:
filter_id is the right one. I typed this by mistake because in my stm32 implementation its the same.

tbursztyka commented on include/can.h:
```diff
@@ -0,0 +1,380 @@
+ * @retval -EINVAL if num_bytes > 8.
+ */
+static inline int can_write(struct device *dev, u8_t *data,
+							u8_t num_bytes, u32_t id, enum can_rtr rtr,
```
rename num_bytes to length, it's a common name among APIs

tbursztyka commented on include/can.h:
```diff
@@ -0,0 +1,380 @@
+		msg.id_type = CAN_STANDARD_IDENTIFIER;
+		msg.std_id = id;
+	}
+	msg.dlc = num_bytes;
```
try to break a line after '}' ( unless it's another '}' )

tbursztyka commented on include/can.h:
```diff
@@ -0,0 +1,380 @@
+								 struct can_filter *filter);
+
+/**
+ * @brief Attach a isr callback function to a single or group of identifiers.
```
an isr (no ?)

tbursztyka replied:
no I meant you wrote "a isr", but shouldn't it be "an isr"

The name is just fine actually.

alexanderwachter replied:
I created this function with further work on high level protocols in mind.
This reduces copying of the message and a high level protocol can easily attach himself to this specific identifiers.

alexanderwachter replied:
oh misspelling :-)

tbursztyka commented on drivers/can/Kconfig:
```diff
@@ -0,0 +1,75 @@
+#
+menuconfig CAN
+	bool
+	prompt "CAN Drivers"
```
you can coalesce the type line with the prompt message, it avoids having a dedicated prompt line, it's shorter:
bool "CAN Drivers"

Apply that everwhere in the Kconfig file (you are actually doing it already in some options)

alexanderwachter replied:
Sry. Copy and paste from another driver.

tbursztyka commented on drivers/CMakeLists.txt:
```diff
@@ -26,6 +26,7 @@ add_subdirectory_if_kconfig(spi)
 add_subdirectory_if_kconfig(usb)
 add_subdirectory_if_kconfig(watchdog)
 add_subdirectory_if_kconfig(wifi)
+add_subdirectory_if_kconfig(can)
```
That should go in the previous patch, the one adding the Kconfig infrastructure.

tbursztyka commented on drivers/can/Kconfig.stm32:
```diff
@@ -0,0 +1,25 @@
+
+config CAN_STM32
+	bool
+	prompt "STM32 CAN Driver"
```
Same as previous Kconfig patch: coalesce the type line with the prompt message

tbursztyka commented on drivers/can/stm32_can.c:
```diff
@@ -0,0 +1,765 @@
+
+	__ASSERT_NO_MSG(data != NULL);
+
+	u32_t bus_off = can->ESR & CAN_ESR_BOFF;
```
Always declare a variable at the beginning of its code block.

tbursztyka commented on drivers/can/stm32_can.c:
```diff
@@ -0,0 +1,765 @@
+	const struct can_stm32_config *cfg;
+	CAN_TypeDef *can;
+
+	__ASSERT_NO_MSG(arg != NULL);
```
That's never going to happen, you can get rid of this.

tbursztyka commented on drivers/can/stm32_can.c:
```diff
@@ -0,0 +1,765 @@
+	cfg = DEV_CFG(dev);
+	can = cfg->can;
+
+	__ASSERT_NO_MSG(data != NULL);
```
Unlikely to happen as well

tbursztyka commented on drivers/can/stm32_can.c:
```diff
@@ -0,0 +1,765 @@
+	if (can->TSR & CAN_TSR_TME_Msk) {
+		k_sem_give(&data->tx_int_sem);
+	}
+	while (can->RF0R & CAN_RF0R_FMP0_Msk) {
```
break a line before the while

tbursztyka commented on drivers/can/stm32_can.c:
```diff
@@ -0,0 +1,765 @@
+
+		SYS_LOG_DBG("Message on filter index %d", filter_match_index);
+		can_get_msg_fifo(mbox, &msg);
+		if (data->rx_response[filter_match_index]) {
```
break a line before the if.
The only time you can pack the lines like that is when the if condition is testing previous line's assignment. Works the same for while/for 

Apply that everywhere.

tbursztyka commented on drivers/can/stm32_can.c:
```diff
@@ -0,0 +1,765 @@
+		k_sem_give(&data->tx_int_sem);
+	}
+	while (can->RF0R & CAN_RF0R_FMP0_Msk) {
+		struct can_msg msg;
```
Usually we order variable declaration so the ones whiche are pre-assigned are first, so move this one as last one.

Apply that everywhere

tbursztyka commented on drivers/can/stm32_can.c:
```diff
@@ -0,0 +1,765 @@
+{
+	ARG_UNUSED(hcan);
+}
+static int can_stm32_init(struct device *dev)
```
alway break a line before declaring new function

tbursztyka commented on drivers/can/stm32_can.c:
```diff
@@ -0,0 +1,765 @@
+	hal_mode =  mode == CAN_NORMAL_MODE   ? CAN_MODE_NORMAL   :
+				mode == CAN_LOOPBACK_MODE ? CAN_MODE_LOOPBACK :
+				mode == CAN_SILENT_MODE   ? CAN_MODE_SILENT   :
+											CAN_MODE_SILENT_LOOPBACK;
```
indentation issue. Be aware we enforce 80 chars limit rule.

alexanderwachter replied:
Sorry for that. I accidentally used spaces instead of tabs and uncrustify replaced them a little bit strange.

tbursztyka commented on drivers/can/stm32_can.c:
```diff
@@ -0,0 +1,765 @@
+		     filter_type == CAN_FILTER_EXTENDED_MASKED) ?
+		    "with" : "without",
+		    filter_type);
+	do {
```
hum, that's an ugly code block with too many if/if/if...
Could you rewrite this? You can externalize some logic to inline functions if you want.

 (btw, why not using a switch on filter_type here? As you are doing at line 581)

alexanderwachter replied:
I will try to reduce it to 2 levels.
The switch is not used because of the break.

alexanderwachter replied:
Done. I managed to get rid of the ifs acting like a switch.

tbursztyka commented on drivers/can/stm32_can.c:
```diff
@@ -0,0 +1,765 @@
+	enum can_filter_type old_filter_type = can_get_filter_type(bank_bit,
+														  can->FM1R, can->FS1R);
+	/* TODO fifo balancing */
+	if (filter_type != old_filter_type) {
```
you could rewrite this code block to gain in clarity as well. Externalizing the logic into a dedicated inline function.

alexanderwachter replied:
Agree!

tbursztyka commented on drivers/can/stm32_can.c:
```diff
@@ -0,0 +1,765 @@
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_CAN_LEVEL
+#include <logging/sys_log.h>
+
+static const struct can_driver_api can_api_funcs = {
```
We always declare the driver API at the end of the driver code, before the DEVICE_INIT() macro call.

Follow that, and then remove the function declarations in stm32_can.h

tbursztyka commented on samples/can/src/main.c:
```diff
@@ -0,0 +1,336 @@
+			continue;
+		}
+		switch (msg.data[0]) {
+		case 0:
```
Could you use macros that would self-describe the case?

alexanderwachter replied:
Sure.

tbursztyka reviewed (changes requested):
Lost of style issue. Make sure to run checkpatch script as documented here: http://docs.zephyrproject.org/contribute/contribute_guidelines.html#coding-style

Btw, how this CAN API would fit with a (hypothetical) future canopen subsystem?

alexanderwachter commented:
@tbursztyka: Yes, I had high level protocols such as iso-tp and canopen in mind. They can attach with attach_isr and copy the messages to the desired buffers/netbuffers/net interfaces without the copy overhead whey would have with msgq.

carlescufi commented:
@jukkar @tbursztyka @erwango this was reworked by the author, can you take another look please?

carlescufi reviewed (approved):
This looks solid to me. I have no direct objections with the API proposed either. It needs rebasing but the changes requested seem to have been addressed. I would propose we include this in 1.12

tbursztyka commented:
Getting this in 1.12 would be nice indeed

jukkar commented on samples/can/Kconfig:
```diff
@@ -0,0 +1,93 @@
+	string "Name of the CAN device"
+	default "CAN_1"
+	help
+		"Name of the can device used for send an receive."
```
The help texts do not follow zephyr style. Also noticed that some of the help text have mismatch ". Anyway, there is no need to put help text inside "". So you can have it like this:

```
	help
	  Name of the can device used for send an receive.

```

jukkar commented on samples/can/src/main.c:
```diff
@@ -0,0 +1,348 @@
+#include <device.h>
+#include <can.h>
+#include <gpio.h>
+
```
Remove extra empty line

jukkar commented on samples/can/src/main.c:
```diff
@@ -0,0 +1,348 @@
+
+	ret = gpio_pin_configure(gpio_dev, CONFIG_PIN_LED_1, GPIO_DIR_OUT);
+	gpio_pin_write(gpio_dev, CONFIG_PIN_LED_1, 0);
+	#ifdef CONFIG_USE_LED_2
```
Start the #ifdef from the column 0 here and below

jukkar commented on samples/can/src/main.c:
```diff
@@ -0,0 +1,348 @@
+		return;
+	}
+
+	struct can_filter filter = {
```
Declare the variables at the beginning of the function

jukkar commented on samples/can/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+
```
About the commit message of the sample. Does the commit message mean that you have only tested this with stm32f072b_disco, or is it really only working in that specific board. I did not see anything specific for just that board. If the sample should work in other smt32 boards too, then you could say that "Tested on stm32f072b_disco" instead of current text.

Also there should be a readme file for the sample, see examples what to put there for example in networking samples.

jukkar reviewed (changes requested):
Looks good, just minor comments to the sample app.

erwango commented on drivers/can/stm32_can.h:
```diff
@@ -0,0 +1,75 @@
+ *
+ */
+
+#ifndef _STM32_LL_CAN_H_
```
_STM32_CAN_H_

alexanderwachter replied:
Done.

erwango reviewed (changes requested):
Minor remaining comment. Otherwise LGTM

carlescufi commented:
@alexanderwachter can you try to address the comments so we can get this merged in time for 1.12?

alexanderwachter commented:
@carlescufi I'm working on it but I somehow ruined my local git repo. I had Hash conflicts.
I try to fix this an submit as soon as possible.

alexanderwachter commented:
@carlescufi it seems to be ok now but I forgot to rebase after reordering of the commits.
Now it triggered a lot of reviews from the changes below my first commit and the latest master.
Could you fix that somehow?

carlescufi commented:
@alexanderwachter I am not sure what you mean. This looks OK to me, GitHub doesn't handle rewriting history so it sort of "hides" the comments that are no longer applicable to the current commit set.

pfalcon commented on samples/can/Kconfig:
```diff
@@ -0,0 +1,93 @@
+	help
+	  Pin number of the first LED.
+
+config USE_LED_2
```
I don't see how option named "USE_LED_2" can have a description of "Use two LEDs". "USE_LED_2" clearly means "use LED no.2". Please update option names for better clarity.

Actually, I don't see why boolean options are used. If user can select 1-4 LEDs, then it should be an integer option with the given range.

But the best of all would be following: if this is the basic CAN example, which can work with just 1 LED, please just make it so - the source code will be much more clear (important, because many people will use it as a reference for CAN things), and it will work on as many boards as possible without extra configuration efforts.

Thanks.

alexanderwachter replied:
Done.

pfalcon replied:
Thanks!

alexanderwachter commented:
@carlescufi The commits vanished but the reviewers remain. For example mike-scott or jfischer-phytec-iot.

alexanderwachter commented:
BTW I changed the filter parameter from attach_* from
struct can_filter *filter to
const struct can_filter *filter.
This saves memory because there is no need to have those structs in SRAM region.

carlescufi commented:
@alexanderwachter I removed a couple of reviewers but don't worry, you don't need approval from all in order to merge the PR.

dbkinder commented on samples/can/README.rst:
```diff
@@ -0,0 +1,45 @@
+.. _can-sample:
+
+Controller Area Network
+###########
```
Heading underlines must be at least as long as the title itself. Add some more ``#``'s

dbkinder commented on samples/can/README.rst:
```diff
@@ -0,0 +1,45 @@
+Overview
+********
+
+This sample demonstrates how to use the CAN API.
```
On first use, would be better to spell out the acronym:

    This sample demonstrates how to use the Controller Area Network (CAN) API.

dbkinder commented on samples/can/README.rst:
```diff
@@ -0,0 +1,45 @@
+********
+
+This sample demonstrates how to use the CAN API.
+Messages with standard and extended identifiers are sent over the bus triggered
```
add a comma after ``bus``

dbkinder commented on samples/can/README.rst:
```diff
@@ -0,0 +1,45 @@
+This sample demonstrates how to use the CAN API.
+Messages with standard and extended identifiers are sent over the bus triggered
+by a button event.
+Messages are received using message queues and ISRs. Those messages blink LEDs
```
To avoid the repetition, how about saying:

    Messages are received using message queues and ISRs, and 
    message reception is indicated by blinking LEDs and output
    on the console.

dbkinder commented on samples/can/README.rst:
```diff
@@ -0,0 +1,45 @@
+In loopback mode, the board receives its own messages. This could be used for
+standalone testing.
+
+The sample can be built and executed on boards supporting CAN.
```
How do we know if the board supports CAN?  Better would be to state the hardware requirements needed on the board.

jukkar reviewed (approved):
LGTM

alexanderwachter commented on samples/can/README.rst:
```diff
@@ -0,0 +1,45 @@
+.. _can-sample:
```
I think net is not the right place for it because the samples there are more application layer related.

alexanderwachter replied:
Maybe in future we get high level can protocols and for that we will need a can category anyway.

carlescufi commented:
@alexanderwachter I see you have an error from the documentation build, where the README from CAN is not being included.

alexanderwachter commented:
There is an ongoing discussion about the samples directory. I will include it after we agree on that.

carlescufi commented:
@dbkinder can you please re-review this?
@alexanderwachter can you fix the CI issues?

alexanderwachter commented:
@carlescufi should be ok now. Something went wrong with the merge.
